### PR TITLE
Add support unary and ternary operations

### PR DESCRIPTION
**Pull request by rdspring1**
This PR adds support for unary and ternary operations, so `OpRecord` is completely implemented.

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

File: tests/python/test_python_frontend.py [hunk 1]
```diff
@@ -1098,8 +1098,10 @@ def fusion_func(fd: FusionDefinition):
         assert n.dtype == torch.complex64
 
     def test_where_op(self):
+        # nvfuser_where is a decorator function. It takes the input arguments
```
Comment (rdspring1) [inline tests/python/test_python_frontend.py:1101]: I refactored `test_where_op` to be compatible with `exec_nvfuser`.

**Comment by rdspring1** [comment]
!build

File: csrc/python_frontend/translation_utils.cpp [hunk 2]
```diff
@@ -104,4 +213,64 @@ std::string getString(const BinaryOp* bop) {
...
+            false,                                                           \
+            "Unexpected operator type: ",                                    \
+            top->getTernaryOpType(),                                         \
+            " in ",                                                          \
+            top->toString());                                                \
+    }                                                                        \
+  }
+
+// Template specializations for std::function for TernaryOp
+GET_FUNCTION_TERNARY_SPECIALIZATION(TensorView*, TensorView*, Val*, Val*)
```
Comment (jacobhinkle) [inline csrc/python_frontend/translation_utils.cpp:273]: Are the other combinations needed here, like TV,TV,Val and Val,TV,Val etc?

**Review by jacobhinkle — APPROVED** [review]

File: csrc/python_frontend/translation_utils.cpp (hunk 2 above)
Comment (jjsjann123) [reply csrc/python_frontend/translation_utils.cpp:273]: These are just for `clamp` & `threshold`. So I don't think we needed anything else.

Comment (jjsjann123) [reply csrc/python_frontend/translation_utils.cpp:273]: I think moving this closer to where the generic template is would help clearing the confusion.

File: csrc/python_frontend/translation_utils.cpp [hunk 3]
```diff
@@ -104,4 +213,64 @@ std::string getString(const BinaryOp* bop) {
...
+GET_FUNCTION_TERNARY_SPECIALIZATION(Val*, Val*, Val*, Val*)
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply csrc/python_frontend/translation_utils.cpp:274]: The template is a fully specified function, so it violates the `One Definition Rule` if it is placed in the header.

Linker error message:
```
/usr/bin/ld: CMakeFiles/codegen_internal.dir/csrc/python_frontend/translation_utils.cpp.o: in function `std::function<nvfuser::TensorView* (nvfuser::TensorView*, nvfuser::Val*, nvfuser::Val*)> nvfuser::python_frontend::getFunction<nvfuser::TensorView*, nvfuser::TensorView*, nvfuser::Val*, nvfuser::Val*>(nvfuser::TernaryOp const*)':
translation_utils.cpp:(.text+0xc30): multiple definition of `std::function<nvfuser::TensorView* (nvfuser::TensorView*, nvfuser::Val*, nvfuser::Val*)> nvfuser::python_frontend::getFunction<nvfuser::TensorView*, nvfuser::TensorView*, nvfuser::Val*, nvfuser::Val*>(nvfuser::TernaryOp const*)'; CMakeFiles/codegen_internal.dir/csrc/python_frontend/translation.cpp.o:translation.cpp:(.text+0x9d0): first defined here
```

File: csrc/python_frontend/translation_utils.h [hunk 4]
```diff
@@ -108,7 +220,37 @@ std::function<ResultType(ArgTypes...)> getFunction(const BinaryOp* bop) {
...
+    case TernaryOpType::Lerp:
+      return wrap_function(lerp);
+      break;
+    case TernaryOpType::Where:
+      return wrap_function(where);
+      break;
+    default:
+      NVF_CHECK(
+          false,
+          "Unexpected operator type: ",
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply csrc/python_frontend/translation_utils.h:240]: I added error messages for `clamp` and `threshold` in the general template function.

File: csrc/python_frontend/translation_utils.cpp (hunk 2 above)
Comment (rdspring1) [reply csrc/python_frontend/translation_utils.cpp:274]: Since the template functions are fully specialized, they must go in the cpp file.

**Comment by rdspring1** [comment]
!build

File: csrc/python_frontend/translation_utils.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/python_frontend/translation_utils.cpp:275]: good call. And I realized that you are using macro here as well.

I'm a little bit paranoid of having a full template function specialization hidden inside a translation unit.
Alternative 1. If this isn't a perf critical piece, I think we can get away with having it inlined.
Alternative 2. If we don't mind adding some extra macro on the header, we can have the specialization declared in the header while keeping the definition in the cpp file.

[the remark this answers is not in the document]
Comment (rdspring1) [reply csrc/python_frontend/translation_utils.cpp]: I went with alternative 2. Declared the specialization in header without macro. The definition is created in the cpp file with macro.

```
// Fully specialized template functions to create std::function for TernaryOp.
template <>
std::function<TensorView*(TensorView*, Val*, Val*)> getFunction<
    TensorView*,
    TensorView*,
    Val*,
    Val*>(const TernaryOp* top);

template <>
std::function<Val*(Val*, Val*, Val*)> getFunction<Val*, Val*, Val*, Val*>(
    const TernaryOp* top);
```

[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/python_frontend/translation_utils.cpp]: Awesome. Thanks a lot for bearing with my nitpick 🙇

**Pull request merged**